src/realm/table.cpp

```diff
@@ -1080,69 +1080,116 @@ void Table::set_table_type(Type table_type)
+                auto value = leaf.get_as_ref_or_tagged(i);
+                if (value.is_ref() && value.get_as_ref() == 0) {
+                    // ref of zero means there's no backlinks
+                    continue;
+                }
+
+                if (value.is_ref()) {
+                    // Any other ref indicates an array of backlinks, which will
+                    // always have more than one entry
+                    incoming_link_count[i] = LinkCount::Multiple;
+                }
+                else {
+                    // Otherwise it's a tagged ref to the single linking object
+                    if (incoming_link_count[i] == LinkCount::None) {
+                        incoming_link_count[i] = LinkCount::One;
+                    }
+                    else if (incoming_link_count[i] == LinkCount::One) {
+                        incoming_link_count[i] = LinkCount::Multiple;
                     }
                 }
-                backlink_count += cur_backlinks;
-                if (backlink_count > 1) {
-                    throw std::logic_error(
-                        util::format("At least one object in '%1' does have multiple backlinks.", get_name()));
+
+                auto source_col = get_opposite_column(col);
+                if (source_col.get_type() == col_type_Mixed) {
+                    auto source_table = get_opposite_table(col);
+                    throw std::logic_error(util::format(
+                        "Cannot convert '%1' to embedded: there is an incoming link from the Mixed property '%2.%3', "
+                        "which does not support linking to embedded objects.",
+                        get_class_name(), source_table->get_class_name(), source_table->get_column_name(source_col)));
                 }
-                return IteratorControl::AdvanceToNext; // continue
-            });
+            }
+            return IteratorControl::AdvanceToNext;
+        });
 
-            if (backlink_count == 0) {
-                throw std::logic_error(util::format(
-                    "At least one object in '%1' does not have a backlink (data would get lost).", get_name()));
+        for (size_t i = 0; i < size; ++i) {
+            if (incoming_link_count[i] == LinkCount::None) {
+                orphans.push_back(cluster->get_real_key(i));
             }
+            else if (incoming_link_count[i] == LinkCount::Multiple) {
+                multiple_incoming_links.push_back(cluster->get_real_key(i));
+            }
+        }
+
+        return IteratorControl::AdvanceToNext;
+    });
+
+    if (handle_backlinks) {
+        for (auto key : orphans) {
+            remove_object(key);
+        }
+        for (auto key : multiple_incoming_links) {
+            auto obj = get_object(key);
+            obj.handle_multiple_backlinks_during_schema_migration();
+            obj.remove();
+        }
+    }
+    else {
+        if (!orphans.empty()) {
```

Comment:
We don't need to have the entire list populated to throw these exceptions, can we move this check on `!handle_backlinks` into the traversal code so that it stops immediately on the first one found?